stdlib/public/runtime/SwiftRT-ELF.cpp

```diff
@@ -85,4 +81,4 @@ struct {
   const char MagicString[sizeof(SWIFT_REFLECTION_METADATA_ELF_NOTE_MAGIC_STRING)];
   const swift::MetadataSections *Sections;
 } __attribute__((__packed__))
-Note = {SWIFT_REFLECTION_METADATA_ELF_NOTE_MAGIC_STRING, &sections};
+Note = {SWIFT_REFLECTION_METADATA_ELF_NOTE_MAGIC_STRING, &__swift_metadataSections};
```

Comment:
@jckarter, if I'm not mistaken we still have a runtime relocation here (&__swift_metadataSections will have to be adjusted), but yeah, the total number of relocations has decreased.

Reply:
True, but I think that's OK given that it would have to be filled in with the absolute address during the static constructor anyway when we build the linked list.

Reply:
@jckarter - I'm wondering if it would be possible to use a similar approach to what you did for MetadataSections  (switch to using RelativePointer) / maybe I'm missing something ?

Reply:
The linked list nodes have to be able to reach across dynamic libraries in order to link together the sections in each dynamic library. These links fundamentally aren't known until runtime when the dynamic linker loads the process.

Reply:
@jckarter, right, but I meant the following: The symbol __swift_metadataSections is hidden, so e.g. we can have one ctor (to avoid issues with out-of-order calls of global ctors), and among other things it will populate the struct Note (similarly to your code above) and then call the registration code  etc.  Am I wrong ?

Reply:
We could in principle make the pointer to the local `__swift_metadataSections` structure relative, yeah. Since the page is going to get dirtied during load time anyway, and it wouldn't make the structure much smaller, it didn't seem like a win to do so. The side costs of relocation are higher on ELF, though.

Reply:
i see, so basically i was thinking about the issues with the previous attempt, one of them - the linker might emit different dynamic relocations => the code in swift-reflection-dump.cpp might get complicated because of this. If we didn't have to deal with relocations then we would only have to "slide" all the allocatable sections and avoid modifying their content.

Reply:
My thinking was that the note and `swift-reflection-dump` could point directly at `__swift_metadataSections` within the various images. This linked list data structure is mostly for the runtime's benefit.

Reply:
khm, interesting. When we run swift-reflection-dump -binary-filename a.out, there are several things which i was keeping in mind: (A) a.out can be a statically linked (ELF) executable (on Linux), (B) a.out can be a dynamic library (.so) (ELF), (C) the host platform where we run this test potentially can be different from Linux, e.g. OSX or Windows. All together, (A), (B), (C) quite limit the possibilities what we potentially can do inside swift-reflection-dump with the input file a.out , i.e. because of (A) and (C) we can't dlopen it. Next, in general the bytes dedicated for the field Sections (of the struct Note) might be zero, might be non-zero (depending on how a.out was built), and, again, depending on how a.out was built we might have (or not have) a dynamic relocation which tells the dynamic loader how to patch those bytes to make the field Sections (of the struct Note) actually contain the correct valid pointer to __swift_metadataSections. These were my thoughts, but maybe I'm missing smth. Basically just curious how different pieces will work together here.

Reply:
@jckarter - ok, is there anything blocking this PR ? (if this gets committed then we can try to fix parsing of ELF images)

Reply:
It seems like the pointer to `& __swift_metadataSections` from the note could be relativized. I don't think there's anything blocking this now that we have a workaround for the lld issue. We'd have to work through the tests and fix any other places that happen to be hardcoding the existing data structure.

Reply:
@jckarter Thanks!